From 25ca6deb51bf9da03e95dfcb133390cbbe3cac2d Mon Sep 17 00:00:00 2001
From: Kazunori Kobayashi <kkobayas@igel.co.jp>
Date: Wed, 1 Aug 2012 10:50:50 +0900
Subject: [PATCH 45/97] ext/dfbvideosink: post a message that means frames
 begin to be rendered.

This message is posted for the GStreamer application.
---
 ext/directfb/dfbvideosink.c |   11 +++++++++++
 ext/directfb/dfbvideosink.h |    1 +
 2 files changed, 12 insertions(+)

diff --git a/ext/directfb/dfbvideosink.c b/ext/directfb/dfbvideosink.c
index adb6238..54856e9 100644
--- a/ext/directfb/dfbvideosink.c
+++ b/ext/directfb/dfbvideosink.c
@@ -1610,6 +1610,7 @@ gst_dfbvideosink_change_state (GstElement * element, GstStateChange transition)
       }
       break;
     case GST_STATE_CHANGE_PAUSED_TO_PLAYING:
+      dfbvideosink->frame_rendered = TRUE;
       break;
     default:
       break;
@@ -1621,6 +1622,7 @@ gst_dfbvideosink_change_state (GstElement * element, GstStateChange transition)
 
   switch (transition) {
     case GST_STATE_CHANGE_PLAYING_TO_PAUSED:
+      dfbvideosink->frame_rendered = FALSE;
       break;
     case GST_STATE_CHANGE_PAUSED_TO_READY:
       dfbvideosink->fps_d = 0;
@@ -1999,6 +2001,14 @@ gst_dfbvideosink_show_frame (GstBaseSink * bsink, GstBuffer * buf)
     }
   }
 
+  if (dfbvideosink->frame_rendered) {
+    gst_element_post_message (GST_ELEMENT_CAST (dfbvideosink),
+        gst_message_new_element (GST_OBJECT_CAST (dfbvideosink),
+            gst_structure_new ("FrameRendered",
+                "message", G_TYPE_STRING, "First frame was rendered", NULL)));
+    dfbvideosink->frame_rendered = FALSE;
+  }
+
 beach:
   return ret;
 }
@@ -2667,6 +2677,7 @@ gst_dfbvideosink_init (GstDfbVideoSink * dfbvideosink)
   dfbvideosink->vsync = TRUE;
   dfbvideosink->setup = FALSE;
   dfbvideosink->running = FALSE;
+  dfbvideosink->frame_rendered = FALSE;
 
   dfbvideosink->cb_channels = NULL;
   dfbvideosink->brightness = -1;
diff --git a/ext/directfb/dfbvideosink.h b/ext/directfb/dfbvideosink.h
index d5b4482..5a1ce39 100644
--- a/ext/directfb/dfbvideosink.h
+++ b/ext/directfb/dfbvideosink.h
@@ -112,6 +112,7 @@ struct _GstDfbVideoSink {
   gboolean vsync;
   gboolean setup;
   gboolean running;
+  gboolean frame_rendered;
   
   /* Color balance */
   GList *cb_channels;
-- 
1.7.10.4

